tests/optim/test_oss.py

```diff
@@ -334,3 +334,62 @@ def test_collect_shards():
     mp.spawn(
         run_test_collect_shards, args=(world_size, reference_rank), nprocs=world_size, join=True,
     )
+
+
+def run_test_multiple_groups(rank, world_size):
+    # Only work with the even ranks, to check that the global_rank indexing is properly used
+    dist_init(rank, world_size)
+    sub_group_ranks = [0, 2, 4]
+    pg = torch.distributed.new_group(ranks=sub_group_ranks, backend="gloo")
+    device = "cpu"
+    epochs, batch, input_width, hidden, target_width = 5, 3, 20, 10, 5
+    loss_fn = torch.nn.L1Loss().to(device)
+
+    def check(optimizer):
+        # Just run a couple of random epochs, check that the model is properly updated
+        for _ in range(epochs):
+            target = torch.rand((batch, target_width), device=device)
+            inputs = torch.rand((batch, input_width), device=device)
+
+            def closure():
+                optimizer.zero_grad()
+                output = model(inputs)
+                loss = loss_fn(output, target)
+                loss.backward()
+
+                return loss
+
+            original_loss = optimizer.step(closure=closure)
+            post_update_loss = loss_fn(model(inputs), target)
+
+            assert (
+                original_loss - post_update_loss
+            ).norm() > 1e-6, f"Model update failed - difference {(original_loss - post_update_loss).norm()}"
+
+    # Model fitting in the broadcast bucket
+    if rank in sub_group_ranks:
+        model = torch.nn.Sequential(torch.nn.Linear(input_width, hidden), torch.nn.Linear(hidden, target_width)).to(
+            device
+        )
+
+        # With SGD, Momentum is required to get a state to shard
+        optimizer = optim.OSS(model.parameters(), lr=0.1, momentum=0.99, group=pg, broadcast_buffer_size=2 ** 20)
+        check(optimizer)
+
+    # Model not-fitting in the broadcast bucket
+    if rank in sub_group_ranks:
```

Comment:
This if seems redundant. Why not merge into the if above?

Reply:
that was an afterthought, can be cleaned up indeed